docs/notebooks/openai_gym_lunar_lander.pct.py

```diff
@@ -0,0 +1,244 @@
+
+# %%
+N_RUNS = 10
+def lander_objective(x):
+    # for each point compute average reward over n_runs runs
+    all_rewards = []
+    for w in x.numpy():
+        rewards = [
+            demo_heuristic_lander(env, w)
+            for _ in range(N_RUNS)
+        ]
+        all_rewards.append(rewards)
+
+    rewards_tensor = tf.convert_to_tensor(all_rewards, dtype=tf.float64)
+    
+    # triste minimizes, and we want to maximize
+    return -1 * tf.reshape(tf.math.reduce_mean(rewards_tensor, axis=1), (-1, 1))
+
+observer = trieste.objectives.utils.mk_observer(lander_objective)
+
+# %% [markdown]
+# ### Solving the optimization problem with Trieste
+#
+# Here we do normal steps required to solve an optimization problem with Trieste: generate some initial data, create a surrogate model, define an acquisition funciton and rule, and run the optimization. Optimization step may take a few minutes to complete.
+
+# %%
+num_initial_points = 2 * search_space.dimension
+initial_query_points = search_space.sample(num_initial_points)
+initial_data = observer(initial_query_points)
+
+
+# %%
+def build_model(data):
+    variance = tf.math.reduce_variance(data.observations)
+    kernel = gpflow.kernels.RBF(variance=variance)
+    gpr = gpflow.models.GPR(data.astuple(), kernel)
+    gpflow.set_trainable(gpr.likelihood, False)
+    # Since we are running multiple simulations per observation,
+    # it is possible to estimate likelihood variance from this data for each observation.
+    # This is possible to model with VGP, as described here:
+    # https://gpflow.readthedocs.io/en/master/notebooks/advanced/varying_noise.html\
+    # In the interest of brevity we have chosen not to do it in this notebook.
+    return trieste.models.gpflow.GaussianProcessRegression(gpr)
+
+model = build_model(initial_data)
+
+# %%
+acq_fn = trieste.acquisition.function.ExpectedImprovement()
+rule = trieste.acquisition.rule.EfficientGlobalOptimization(acq_fn)  # type: ignore
+
+# %%
+N_OPTIMIZATION_STEPS = 200
+bo = trieste.bayesian_optimizer.BayesianOptimizer(observer, search_space)
+result = bo.optimize(N_OPTIMIZATION_STEPS, initial_data, model, rule).final_result.unwrap()
+
+# %% [markdown]
+# ### Analyzing the results
+#
+# First, let's just plot a regular regret plot of our optimization, to ensure Trieste indeed found a better configuration of the controller. Remember that we flipped the sign of the reward.
+
+# %%
+from util import plotting
+import matplotlib.pyplot as plt
+
+ax = plt.gca()
+plotting.plot_regret(
+    result.dataset.observations.numpy(),
+    ax,
+    num_init=len(initial_data)
+)
+
+# %% [markdown]
+# We can also retrieve the best configuration found, and test it. We expect to see mostly large positive rewards here.
+
+# %%
+w_best = result.dataset.query_points[np.argmin(result.dataset.observations), :]
```

Comment:
this makes me a bit sad... we should take the argmin of the GPR prediction at dataset.query_points. With a bit of luck they are the same!

Reply:
please don't be sad, Victor! they are the same indeed, so i've changed it. hope it makes you happy!

Reply:
how about adding a sentence explaining our choice? e.g. "We choose here the query point that gives the best predictive expected reward according to our model".

Reply:
souds good!